Feature/persistance service

Heffertron opened the pull request:
PR with latest commits. Happy to discuss next time we chat.

Heffertron committed:
- Imported RealmSwift library
- Objects now 'final' classes rather than structs to accommodate Realm
- Created PersistenceService.save for saving questions from Firebase to Realm
- Created PersistanceService.getQuestions for retrieving questions from Realm
- Created a Realm extension file

Heffertron committed:
- Made some changes to Persistence saving to make using Realm safer using a do, try, catch
- Struggling with saving an array of Int to Realm - pushed for Ben to review

BenSullivan-iOS committed:
Fixing realm save issues (#3)

Heffertron committed:
- PersistenceService made using Realm safer
- Question: added a computed property to pull out the correct answer
- User object added code for the two methods that Ben started
- View controller - added a temporary button that allows us to check that getting the questions from Realm works as well as using the computed property of finding a correct answer

Heffertron committed:
Latest build of the persistence-service, changes since last commit:
- Removed my for-loop on computed property in Question object in favour of Ben's functional programming magic
- Added logic for method 'removeAnsweredQuestion' in the User object - need to discuss this further
- Added two buttons to the ViewController, one to fetch all questions from Realm, the other to remove questions - this is temporary and was primarily for me to check some of the methods worked

benskill commented on Share The Load/Model/User.swift:
```diff
@@ -0,0 +1,43 @@
+    
+    /// Had a question on this in our Twitter regarding whether we should be 'removing' the Question object from Realm or just filtering out a list of id's when we access the data from Realm.
+    ///If we can talk this through on a call, then I can crack on
+    func removeAnsweredQuestion(questionAnswered: Question) {
```
I think using the id rather than the question is still the way to go. I think we can find the question with the relevant id then remove it.

Doing this means the User object is in no way dependant on the Question object. Knowing less about other models is always a good thing.

benskill commented on Share The Load/Screens/ViewController.swift:
```diff
@@ -0,0 +1,102 @@
+        createRemoveButton()
+    }
+    
+    
```
To be really picky, can we just leave one empty line between funcs 😬

Heffertron replied:
Hit the easy one first, done. 👍🏼

benskill commented on Share The Load/Screens/ViewController.swift:
```diff
@@ -0,0 +1,102 @@
+        view.addSubview(getRealmQuestionsButton)
+        getRealmQuestionsButton.translatesAutoresizingMaskIntoConstraints = false
+        getRealmQuestionsButton.setTitle("Get Questions From Realm", for: .normal)
+        getRealmQuestionsButton.frame = view.frame
```
Shouldn't need to do this if you're using constraints below

Heffertron replied:
Got it, removed.

benskill commented on Share The Load/Screens/ViewController.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+    
+    
+    @objc func buttonTapped() {
```
Which button?

Heffertron replied:
Yep - I see it. Thanks - changed to: 'getRealmQuestionsButtonTapped'.

benskill commented on Share The Load/Screens/ViewController.swift:
```diff
@@ -0,0 +1,102 @@
+        
+        print("Questions from Realm before removal: \(questionsFromRealm[0])")
+        
+        let user = User()
```
Maybe make this user global for the time being before we figure out how rammed it properly accessible. That way we can use it to test adding and removing various questions

benskill replied:
Looks like we already have a global one you can use

Heffertron replied:
I've remove this completely. This was based on the User object removing a Question object directly from Realm, which we've agreed isn't the best way to do this. So just deleting this code and going to implement the logic once I've updated the User object.

benskill commented on Share The Load/Services/PersistanceService.swift:
```diff
@@ -0,0 +1,70 @@
+/// Massive temporary hack to get save logic working
+/// We'll need a UserService to always give us the same instance of a user rather than this global properties
+/// Global properties are very bad.
+var user = User()
```
This is the global one I was referring to

Heffertron replied:
Same as above comment. Not using anymore.

benskill commented on Share The Load/Services/PersistanceService.swift:
```diff
@@ -0,0 +1,70 @@
+    
+    
+    func getQuestions() -> [Question] {
+        var questionsFromRealm = [Question()]
```
This initialises an array with an empty question as the first element. We ideally want an empty array so you can do var questionsFromRealm: [Question] = []

Heffertron replied:
Sounds good - updated.

benskill commented on Share The Load/Screens/ViewController.swift:
```diff
@@ -0,0 +1,102 @@
+class ViewController: UIViewController {
+    
+    private let getRealmQuestionsButton = UIButton()
+    private let removeFromRealmButton = UIButton()
```
Remember the concept of Realm shouldn't be something any other objects are aware of. We designed the PersistenceService public interface with either objects don't need to know where it gets its data from.

Heffertron replied:
Are you suggesting that these buttons should be renamed so that they don't include a reference to Realm?

Heffertron replied:
FYI - updated in last commit as per your DM on Twitter

benskill commented on Share The Load/Model/Quiz.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+final class Quiz: Codable {
```
I don't think we'll need this codable unless we want it sent from the backend

benskill replied:
I think this object should be our next big task. Have a think about what it should do and we can discuss

Heffertron replied:
Removed codable, and will start to have a think about this object.

benskill commented on Share The Load/Screens/ViewController.swift:
```diff
@@ -0,0 +1,102 @@
+
+class ViewController: UIViewController {
+    
+    private let getRealmQuestionsButton = UIButton()
```
Try this syntax for views, when the button is first accessed (usually to add it to the view) this closure is called which returns the setup button. It's only called once, from then on it's just the button that it originally returned 

Private var button: UIButton =  {
Let button = UIButton()
//configure button as needed but don't do the layout here

return button
}))

Heffertron replied:
Thanks for your help on this this evening. I do have one question about what is deemed 'layout' exactly...? 

For example, when the closure is called you'll see I've included setting the title, background colour and then adding the target. 

I've then called the corner radius and the TAMIC in the layout method. 

Is that accurate? Or should I have put the corner radius and TAMIC in the button closure?

benskill replied:
The corner radius can only be set once the view has laid out, otherwise the height could be incorrect. It ideally needs to be called in viewDidLayoutSubviews so we know it's actually been laid out

benskill commented on Share The Load/Model/Question.swift:
```diff
@@ -0,0 +1,22 @@
+    var answers = List<Answer>()
+    
+    var correctAnswer: Answer {
+        answers.first { $0.id == 1 }!
```
Probably a good time to come back to this force unwrap. We're saying we want the app to crash if any question doesn't have an answer. I think we should change this to return an optional instead. Then we can easily filter out ones with potentially nil answers when it's accessed

benskill commented on Share The Load/Screens/ViewController.swift:
```diff
@@ -0,0 +1,102 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        fetchQuiz()
+        createGetRealmQuestionsButton()
```
For setting up constraint based layouts you can use loadView() instead of viewDidLoad, it's worth having a look at where that fits in the VC lifecycle, we use it quite a lot on confirmed

Heffertron replied:
Good to know - I've not used loadView() before, so i've implemented this now. Let me know if the implementation of it is 👍🏼 or 👎🏼 on my next PR.

benskill commented on Share The Load/Screens/ViewController.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+    
+    
+    @objc func removeButtonTapped() {
```
It's good to get in the habit of making all these private

Heffertron replied:
Thanks. I've made these private now. So would you recommend that any methods I don't plan to use outside the scope of a file to be private?

benskill replied:
Definitely, the only things which aren't private should be properties or functions which you explicitly want other objects to be able to view or call.

All outlets should be private too, if you have a text label outlet, and you want the text to be set by another object, then you'd create a public string in your class which sets the label's text for you.

Heffertron committed:
Made amends based on comments from Ben on last PR, these include:
- Removing double spaces between functions
- PersistenceService getQuestions() method creates an empty array, rather than an array with a Question object
- Removed codable from Quiz object
- Removed logic from User object 'removeAnsweredQuestion()' method and updated the property to include an Int not a Question
- Made substantial changes to the ViewController mainly around creating the views using loadView()
- Corrected spelling of persistence across whole project

Heffertron committed:
Made changes to the way that the two buttons are created.

Heffertron committed:
Final commit before PR based on Ben's latest feedback, changes in this commit include:
- Made the computed property in the Question object optional for safer use
- Corrected the User object to use an Int rather than a Question so that it has no dependency on the Question object
- In ViewController I added 'if-let' to the use of the computed property in the Question object

Heffertron committed:
Renamed buttons in ViewController to remove reference to Realm

Heffertron commented:
Hey - ready for next review and let me know if changes can be merged.

BenSullivan-iOS reviewed (approved):
Looks good, approved
